include/revng/Pypeline/ObjectID.h

```diff
@@ -0,0 +1,109 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <set>
+
+#include "llvm/ADT/StringRef.h"
+
+#include "revng/Model/TypeDefinition.h"
+#include "revng/Pipeline/Location.h"
+#include "revng/Pipes/Ranks.h"
+#include "revng/Support/MetaAddress.h"
+
+class ObjectID {
+private:
+  using TypeDefinitionKey = model::TypeDefinition::Key;
+  std::variant<std::monostate, MetaAddress, TypeDefinitionKey> Component;
+
+public:
+  enum Kind {
+    Root = 0,
+    Function = 1,
+    TypeDefinition = 2
+  };
+
+public:
+  ObjectID() = default;
+
+  Kind kind() const {
+    if (std::holds_alternative<std::monostate>(Component)) {
+      return Kind::Root;
+    } else if (std::holds_alternative<MetaAddress>(Component)) {
+      return Kind::Function;
+    } else if (std::holds_alternative<TypeDefinitionKey>(Component)) {
+      return Kind::TypeDefinition;
+    } else {
+      revng_abort();
+    }
+  }
+
+  std::vector<std::string> components() const {
+    switch (kind()) {
+    case Kind::Root:
+      return {};
+    case Kind::Function:
+      return { std::get<MetaAddress>(Component).toString() };
+    case Kind::TypeDefinition:
+      return { toString(std::get<TypeDefinitionKey>(Component)) };
+    default:
+      revng_abort();
+    }
+  }
+
+  std::optional<ObjectID> parent() const {
+    switch (kind()) {
+    case Kind::Root:
+      return std::nullopt;
+    case Kind::Function:
+      return ObjectID();
+    case Kind::TypeDefinition:
+      return ObjectID();
+    default:
+      revng_abort();
+    }
+  }
+
+  std::string serialize() const {
+    namespace r = revng::ranks;
+    using pipeline::locationString;
+
+    switch (kind()) {
+    case Kind::Root:
+      return "/root/";
+    case Kind::Function:
+      return locationString(r::Function, std::get<MetaAddress>(Component));
+    case Kind::TypeDefinition:
+      return locationString(r::TypeDefinition,
+                            std::get<TypeDefinitionKey>(Component));
+    default:
+      revng_abort();
+    }
+  }
+
+  bool deserialize(llvm::StringRef Input) {
+    namespace r = revng::ranks;
+    using pipeline::locationFromString;
+
+    if (Input == "/root/") {
+      Component = std::monostate{};
+      return true;
+    } else if (Input.starts_with("/function/")) {
+      auto MaybeComponent = locationFromString(r::Function, Input);
+      if (not MaybeComponent.has_value())
+        return false;
+      Component = std::get<0>(std::get<0>(MaybeComponent->tuple()));
+    } else if (Input.starts_with("/type-definition/")) {
+      auto MaybeComponent = locationFromString(r::TypeDefinition, Input);
+      if (not MaybeComponent.has_value())
+        return false;
+      Component = std::get<0>(MaybeComponent->tuple());
+    }
+    return false;
+  }
+
+  size_t hash() const { return std::hash<decltype(Component)>{}(Component); }
```

Comment:
Maybe I'd line this in the definition of `__hash__`, it's weird in C++.

You can specialize `std::hash<ObjectID>` instead.